src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...newState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+
+      default:
+        break;
```

Comment:
Consider throwing an error or at least logging a warning in the default case of the switch statement. This will help to handle situations where the action type is not recognized, which can be useful for debugging.